contributors/design-proposals/federated-placement-policy.md

```diff
@@ -0,0 +1,271 @@
+
+### 1. Initial Placement Decision
+
+To provide policy-based control over the initial placement decision, we propose
+a new admission controller that integrates with OPA:
+
+When admitting requests, the admission controller executes an HTTP API call
+against OPA. The API call passes the JSON representation of the resource in the
+message body.
+
+The response from OPA contains the desired value for the resource’s annotations
+(defined in policy by the administrator). The admission controller updates the
+annotations on the resource and admits the request:
+
+![InitialPlacement](https://docs.google.com/drawings/d/1c9PBDwjJmdv_qVvPq0sQ8RVeZad91vAN1XT6K9Gz9k8/pub?w=812&h=288)
+
+The update is performed by **merging** the annotations in the response with
+existing annotations on the resource. If there are overlapping keys, the value
+defined by policy is chosen.
+
+#### Example Policy Engine Query:
+
+```http
+POST /v1/data/io/k8s/federation/annotations HTTP/1.1
+Content-Type: application/json
+```
+
+```json
+{
+  "input": {
+    "apiVersion": "extensions/v1beta1",
+    "kind": "ReplicaSet",
+    "metadata": {
+      "annotations": {
+        "eu-jurisdiction-required": "true",
+        "pci-compliance-level": "2"
+      },
+      "creationTimestamp": "2017-01-23T16:25:14Z",
+      "generation": 1,
+      "labels": {
+        "app": "nginx-eu"
+      },
+      "name": "nginx-eu",
+      "namespace": "default",
+      "resourceVersion": "364993",
+      "selfLink": "/apis/extensions/v1beta1/namespaces/default/replicasets/nginx-eu",
+      "uid": "84fab96d-e188-11e6-ac83-0a580a54020e"
+    },
+    "spec": {
+      "replicas": 4,
+      "selector": {...},
+      "template": {...},
+    }
+  }
+}
+```
+
+#### Example Policy Engine Response:
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/json
+```
+
+```json
+{
+  "result": {
+    "federation.kubernetes.io/replica-set-preferences": {
+      "clusters": {
+        "gce-europe-west1": {
+          "weight": 1
+        },
+        "gce-europe-west2": {
+          "weight": 1
+        }
+      },
+      "rebalance": true
+    }
+  }
+}
+```
+
+> This example shows the policy engine returning the replica-set-preferences.
+> The policy engine could similarly return a desired value for other annotations
+> such as the Cluster Selector annotation.
+
+#### Error Handling
+
+The admission controller is designed to **fail closed**. Request handling may
+fail because of:
+
+- Serialization errors
+- Request timeouts or other network errors
+- Unexpected errors from the policy engine
+
+In the event of request timeouts (or other network errors) or back-pressure
+hints from the policy engine, the admission controller should retry after
+applying a backoff.
+
+If the administrator has not defined a policy yet, the response from the policy
+engine indicates no annotations are defined. The admission controller does not
+treat this as an error, i.e., it admits the request.
+
```

Comment:
> Lets say I am an admin and I dont know about the internal implementation. I read about this new policy API resource and I create a new policy. I will expect that all new resources I create should instantly start adhering to these policies. Admission controller should not allow bypassing a defined policy because the policy engine is not running.

This brings up a good point which is that I think any policy resource we add to the API will need to have some "status" recorded on it. This way, policy engines would essentially be controllers (in the Kubernetes sense) for these policy resources. This is important because one of the things we'll want is for the policy to be parsed, compiled, analyzed, etc. before the user is told that it's been accepted. Furthermore, the status could also be made to reflect whether the policy is actually loaded into the policy engine and being **enforced**. So, creation of these policy resources would probably asynchronous, meaning the client would have to check the status to find out whether it's being enforced.

Given the above, we could enable the admission controller by default and rely on the *fail-closed* annotation (mentioned in previous comment) as follows...

1. Initially, the apiserver starts and the admission controller will *fail open*.

2. If at some point, admin toggles the fail-closed annotation to true, then requests would start failing if either (a) policy engine was not deployed or (b) policies have not been loaded into the engine.

3. If the fail-closed annotation is still missing/false, and a policy resource is created, the policy engine would be notified (asynchronously) as usual. The policy engine would parse/compile/install the policy and then call back to the apiserver to automatically set the fail-closed annotation.

We can approximate this as close as possible today by just doing (3) when the policy engine starts and loads the policy via volume mount.

What do you think?